hooks/command

```diff
@@ -1,35 +1,80 @@
 #!/bin/bash
 set -eu -o pipefail
 
+fail() {
+  echo "ERROR: ${*}" 1>&2
+  exit 1
+}
+
+# checks if a corresponding buildkite variable is set,
+# and then if the shell is new enough to support that
+# particular shopt.
+set_shopt() {
+  local option="${1}"
+
+  # `tr` is used instead of Bash 4's ^^ (uppercase) and ,, (lowercase)
+  # because evaluating which version of Bash the agent is running only
+  # happens when setting `shopt` values; it leads to clunkier, more
+  # verbose assignments but it should ensure wider compatability in situations
+  # where no `shopt` values need to be set at all.
+  local varname
+  varname="$(tr '[:lower:]' '[:upper:]' <<< "BUILDKITE_PLUGIN_SHELLCHECK_${option}")"
+  local value
+  value="$(tr '[:upper:]' '[:lower:]' <<< "${!varname:-false}")"
+
+  if [[ ${value} =~ ^(true|1)$ ]]; then
+    bash_version_check "${option}"
+    shopt -s "${option}"
+  fi
+}
+
+bash_version_check() {
+  local option="${1}"
+  local required
+
+  case "${option}" in
+    globstar) required="4" ;;
+    extglob) required="3" ;;
+    *)  fail "attempted to set unknown shopt '${option}'" ;;
+  esac
+
+  if [[ ${BASH_VERSINFO[0]} -lt ${required} ]]; then
+    fail "shopt '${option}' requires Bash ${required} or newer"
+  fi
```

Comment:
Can you add tests for each option to verify the behavior when the BASH_VERSINFO is not enough?

Reply:
`BASH_VERSINFO` is an internal bash variable and I'm not sure it can be overriden. Might require experimentation and/or getting "creative".

Reply:
Yeah, internal bash variables are read-only so this will require shimming in an override if we want to test this:

```
$ bash -x ./test.sh
+ BASH_VERSINFO=(1 2 3 4)
./test.sh: line 3: BASH_VERSINFO: readonly variable
+ echo '5 1 16 1 release x86_64-apple-darwin20.6.0'
5 1 16 1 release x86_64-apple-darwin20.6.0
```

Reply:
OK -- so thinking about this for a few minutes, I think the paths for testing this boil down to:

- Stub in version assignment to allow overriding `BASH_VERSINFO` with a differently named debugging variable during this specific function during testing; effective, but kind of clunky and risky since that override variable can also be set during CI runs (which, like, why would anyone do that? I don't know but it'd be _possible_).
- Modify the `buildkite/plugin-tester` testing container that y'all maintain to allow for multiple versions for Bash to be run. A neat idea, but beginning to approach "ocean boiling" levels of effort.
- Evaluate it locally because it's a pretty straightforward bit of logic comparing an internal bash variable that will only ever be set directly by the actual shell that the plugin is running in (and never, ever inherited because it _can't_ be).

```
$ /bin/bash --version
GNU bash, version 3.2.57(1)-release (x86_64-apple-darwin21)
Copyright (C) 2007 Free Software Foundation, Inc.
$ /usr/local/bin/bash --version
GNU bash, version 5.1.16(1)-release (x86_64-apple-darwin20.6.0)
Copyright (C) 2020 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>

This is free software; you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
$ BUILDKITE_PLUGIN_SHELLCHECK_GLOBSTAR=true BUILDKITE_PLUGIN_SHELLCHECK_FILES="**/*.sh" /bin/bash hooks/command
ERROR: shopt 'globstar' requires Bash 4 or newer
$ BUILDKITE_PLUGIN_SHELLCHECK_GLOBSTAR=true BUILDKITE_PLUGIN_SHELLCHECK_FILES="**/*.sh" /usr/local/bin/bash hooks/command
+++ Running shellcheck on 4 files

In tests/testdata/subdir/llamas.sh line 2:
echo $UNDEFINED_VAR
     ^------------^ SC2086 (info): Double quote to prevent globbing and word splitting.

Did you mean:
echo "$UNDEFINED_VAR"

For more information:
  https://www.shellcheck.net/wiki/SC2086 -- Double quote to prevent globbing ...
$
```

How would you like to proceed?

Reply:
OK, there's a 4th option that occurs to me and is decoupled from the version check: don't check the version, just attempt to set the `shopt` and gate on the success/failure of that:

```
$ shopt -s globstar
-bash: shopt: globstar: invalid shell option name
$ shopt -s globstar &>/dev/null; echo $?
1
```

Reply:
This is fantastic @mckern! Great investigation, thank you very much! 🙏🏻 You are right, and this is practically impossible to test 😅. Your 4th option would be fantastic, but because `shopt` it's a builtin I don't think it would be possible.
I think at this time, we could just probably add a comment to that part of the code with the warning to be careful.

Reply:
Just pushed a refactored commit that no longer uses a strict bash version check and includes numerous notes about what's going on under the hood.

```
$ git describe --always --long
ab24db5
$ BUILDKITE_PLUGIN_SHELLCHECK_GLOBSTAR=true BUILDKITE_PLUGIN_SHELLCHECK_FILES="**/*.sh" /usr/local/bin/bash hooks/command
+++ Running shellcheck on 4 files

In tests/testdata/subdir/llamas.sh line 2:
echo $UNDEFINED_VAR
     ^------------^ SC2086 (info): Double quote to prevent globbing and word splitting.

Did you mean:
echo "$UNDEFINED_VAR"

For more information:
  https://www.shellcheck.net/wiki/SC2086 -- Double quote to prevent globbing ...
$ BUILDKITE_PLUGIN_SHELLCHECK_GLOBSTAR=true BUILDKITE_PLUGIN_SHELLCHECK_FILES="**/*.sh" /bin/bash hooks/command
ERROR: failed to set shopt 'globstar: shopt 'globstar' requires Bash 4 or newer
$ docker run -it --rm -v "$PWD:/plugin:ro" buildkite/plugin-tester
run.bats
 ✓ Shellcheck a single file
 ✓ Shellcheck multiple files
 ✓ Shellcheck multiple files using recursive globbing enabled with 'true'
 ✓ Shellcheck multiple files using extended globbing enabled with '1'
 ✓ Recursive globbing fails if starglob is disabled with 'false'
 ✓ Extended globbing fails if extended globbing is disabled with 'FALSE'
 ✓ Extended globbing fails if extended globbing is disabled with '0'
 ✓ Shellcheck a single file with single option
 ✓ Shellcheck a single file with multiple options
 ✓ Shellcheck multiple files with multiple options
 ✓ Shellcheck failure

11 tests, 0 failures

$
```